// hero-design/src/components/Box/constants.tsx
const COLORS_MAP = {
  black: 'black',
  blue: 'blue',
  'blue-dark-30': 'blueDark30',
  'blue-dark-75': 'blueDark75',
  'blue-light-30': 'blueLight30',
  'blue-light-75': 'blueLight75',
  'blue-light-90': 'blueLight90',
  green: 'green',
  'green-dark-30': 'greenDark30',
  'green-dark-75': 'greenDark75',
  'green-light-30': 'greenLight30',
  'green-light-90': 'greenLight90',
  grey: 'grey',
  'grey-dark-30': 'greyDark30',
  'grey-dark-75': 'greyDark75',
  'grey-light-60': 'greyLight60',
  'grey-light-90': 'greyLight90',
  'grotesque-green': 'grotesqueGreen',
  orange: 'orange',
  'orange-dark-30': 'orangeDark30',
  'orange-dark-75': 'orangeDark75',
  'orange-light-30': 'orangeLight30',
  'orange-light-90': 'orangeLight90',
  pink: 'pink',
  'pink-dark-30': 'pinkDark30',
  'pink-dark-75': 'pinkDark75',
  'pink-light-30': 'pinkLight30',
  'pink-light-90': 'pinkLight90',
  purple: 'purple',
  red: 'red',
  'red-dark-30': 'redDark30',
  'red-dark-75': 'redDark75',
  'red-light-30': 'redLight30',
  'red-light-90': 'redLight90',
  smalt: 'smalt',
  'smalt-dark-75': 'smaltDark75',
  'smalt-light-30': 'smaltLight30',
  'smalt-light-45': 'smaltLight45',
  violet: 'violet',
  white: 'white',
  yellow: 'yellow',
  'yellow-dark-75': 'yellowDark75',
} as const;

const FONT_WEIGHTS_MAP = {
  light: 'light',
  regular: 'regular',
  bold: 'bold',
  'semi-bold': 'semiBold',
} as const;

export { COLORS_MAP, FONT_WEIGHTS_MAP };
