// hero-design/src/components/Breadcrumb/StyledBreadcrumb.tsx
import styled from 'styled-components';
import { Link } from 'react-router-dom';

const SingleBreadcrumbContainer = styled.div`
  display: flex;
  font-size: ${({ theme }) => theme.fontSizes.breadcrumb.item};
  font-weight: ${({ theme }) => theme.fontWeights.breadcrumb.item};
  line-height: ${({ theme }) => theme.lineHeights.breadcrumb.item};
  color: ${({ theme }) => theme.colors.breadcrumb.item};
  margin: 0;
  padding: 0;

  > i {
    line-height: ${({ theme }) => theme.lineHeights.breadcrumb.item};
    margin-right: ${({ theme }) => theme.space.breadcrumb.backIconMarginRight};
  }
`;

const BreadcrumbContainer = styled.ul`
  padding: 0;
  margin: 0;
  font-size: ${({ theme }) => theme.fontSizes.breadcrumb.item};
  font-weight: ${({ theme }) => theme.fontWeights.breadcrumb.item};
  line-height: ${({ theme }) => theme.lineHeights.breadcrumb.item};
  list-style: none;
`;

const BreadcrumbWrapper = styled.li`
  display: inline;
  margin: 0;
  padding: 0;
  color: ${({ theme }) => theme.colors.breadcrumb.itemActive};
`;

const BreadcrumbItemSeparator = styled.span`
  margin: ${({ theme }) => theme.space.breadcrumb.separatorMargin};
  padding: 0;
  color: ${({ theme }) => theme.colors.breadcrumb.item};
`;

const ABreadcrumbItem = styled.a`
  text-decoration: none;
  color: ${({ theme }) => theme.colors.breadcrumb.item};
  margin: 0;
  padding: 0;
`;

const LinkBreadcrumbItem = styled(Link)`
  text-decoration: none;
  color: ${({ theme }) => theme.colors.breadcrumb.item};
  margin: 0;
  padding: 0;
`;

export {
  SingleBreadcrumbContainer,
  BreadcrumbContainer,
  BreadcrumbWrapper,
  BreadcrumbItemSeparator,
  ABreadcrumbItem,
  LinkBreadcrumbItem,
};
