// hero-design/src/components/Checkbox/StyledCheckbox.tsx
import styled, { css } from 'styled-components';

const Label = styled.label`
  display: inline-flex;
  position: relative;
  margin: 0;
  padding: 0;
`;

const LabelText = styled.span`
  margin: 0;
  padding: 0;
  padding-left: ${({ theme }) => theme.space.checkbox.labelPaddingLeft};
  font-size: ${({ theme }) => theme.fontSizes.checkbox.default};
  font-weight: ${({ theme }) => theme.fontWeights.checkbox.default};
  line-height: ${({ theme }) => theme.lineHeights.checkbox.default};
  color: ${({ theme }) => theme.colors.checkbox.text};

  &::before {
    content: '';
    position: absolute;
    width: ${({ theme }) => theme.sizes.checkbox.width};
    height: ${({ theme }) => theme.sizes.checkbox.height};
    top: ${({ theme }) => theme.space.checkbox.top};
    left: 0;
    border: ${({ theme }) => theme.borderWidths.checkbox.default} solid
      ${({ theme }) => theme.colors.checkbox.defaultBorder};
    border-radius: ${({ theme }) => theme.radii.checkbox.default};
    background: ${({ theme }) => theme.colors.checkbox.defaultBackground};
    box-sizing: border-box;
  }

  &:hover {
    cursor: pointer;

    &::before {
      border-color: ${({ theme }) => theme.colors.checkbox.checkedBorder};
    }
  }
`;

const HiddenInput = styled.input`
  display: none;
  margin: 0;
  padding: 0;

  &:checked ~ ${LabelText} {
    &::before {
      border-color: ${({ theme }) => theme.colors.checkbox.checkedBorder};
      background: ${({ theme }) => theme.colors.checkbox.checkedBackground};
    }

    &::after {
      content: '';
      position: absolute;
      height: ${({ theme }) => theme.sizes.checkbox.markHeight};
      width: ${({ theme }) => theme.sizes.checkbox.markWidth};
      top: ${({ theme }) => theme.space.checkbox.markTop};
      left: ${({ theme }) => theme.space.checkbox.markLeft};
      border-width: 0 ${({ theme }) => theme.borderWidths.checkbox.default}
        ${({ theme }) => theme.borderWidths.checkbox.default} 0;
      border-color: ${({ theme }) => theme.colors.checkbox.defaultCheckmark};
      border-style: solid;
      transform: translate(-50%, -30%) rotate(45deg);
      box-sizing: border-box;
    }
  }

  &:indeterminate ~ ${LabelText} {
    &::after {
      content: '';
      position: absolute;
      height: ${({ theme }) => theme.sizes.checkbox.indeterminateBoxHeight};
      width: ${({ theme }) => theme.sizes.checkbox.indeterminateBoxWidth};
      top: ${({ theme }) => theme.space.checkbox.indeterminateBoxTop};
      left: ${({ theme }) => theme.space.checkbox.indeterminateBoxLeft};
      background: ${({ theme }) =>
        theme.colors.checkbox.indeterminateBoxBackground};
      border-radius: ${({ theme }) => theme.radii.checkbox.indeterminateBox};
    }
  }

  &:disabled ~ ${LabelText} {
    color: ${({ theme }) => theme.colors.checkbox.disabledText};

    &:hover {
      cursor: not-allowed;
    }

    &::before {
      border-color: ${({ theme }) => theme.colors.checkbox.defaultBorder};
      background: ${({ theme }) => theme.colors.checkbox.disabledBackground};
    }

    &::after {
      border-color: ${({ theme }) => theme.colors.checkbox.disabledCheckmark};
    }
  }
`;

const GroupWrapper = styled.div<{ themeLayout: 'vertical' | 'horizontal' }>`
  display: flex;
  margin: 0;
  padding: 0;

  ${({ themeLayout, theme }) => {
    switch (themeLayout) {
      case 'vertical':
        return css`
          flex-direction: column;
          & > ${Label}:not(:first-child) {
            margin-top: ${theme.space.checkbox.marginTop};
          }
        `;
      case 'horizontal':
        return css`
          & > ${Label}:not(:first-child) {
            margin-left: ${theme.space.checkbox.marginLeft};
          }
        `;
    }
  }};
`;

export { Label, HiddenInput, LabelText, GroupWrapper };
