// hero-design/src/components/Radio/RadioButtonGroup.tsx
import React, { ReactElement } from 'react';

import RadioButton from './RadioButton';
import { StyledRadioButtonGroupWrapper } from './StyledRadio';

import { CommonProps } from '../common';

export interface RadioButtonGroupProps extends CommonProps {
  /**
   * Radio group name, used for form submission, this MUST be unique in a page, otherwise same name radio group will share the same selection.
   */
  name: string;
  /**
   * Change event handler receiving selected radio's value.
   */
  onChange: (value: string | number) => void;
  /**
   * An array of radio options to be selected.
   */
  options: { disabled?: boolean; text: string; value: string | number }[];
  /**
   * Size of button.
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * Radio input value.
   */
  value?: string | number;
  /**
   * Style of checked button.
   */
  variant?: 'filled' | 'outlined';
}

const RadioButtonGroup = ({
  name,
  value,
  onChange,
  options,
  variant = 'filled',
  size = 'medium',
  id,
  className,
  style,
  'data-test-id': dataTestId,
}: RadioButtonGroupProps): ReactElement => {
  const checkedOptIndex = options.findIndex(
    option => option.value === value && option.disabled !== true
  );
  const preCheckedIndex = checkedOptIndex - 1;

  return (
    <StyledRadioButtonGroupWrapper
      preCheckedIndex={preCheckedIndex}
      id={id}
      className={className}
      style={style}
      data-test-id={dataTestId}
    >
      {options.map(option => {
        return (
          <RadioButton
            key={option.value}
            name={name}
            text={option.text}
            value={option.value}
            disabled={option.disabled}
            checked={option.value === value}
            variant={variant}
            size={size}
            onChange={(e): void => {
              if (e.target.checked === true) {
                onChange(option.value);
              }
            }}
          />
        );
      })}
    </StyledRadioButtonGroupWrapper>
  );
};

export default RadioButtonGroup;
