// hero-design/src/theme/global/colors.ts
const palette = {
  black: '#000000',
  blue: '#1dbeee',
  blueDark30: '#137aa7',
  blueDark75: '#07273c',
  blueLight30: '#5fc7f3',
  blueLight75: '#c6e6fb',
  blueLight90: '#e8f4fd',
  green: '#01b39c',
  greenDark30: '#017d74',
  greenDark75: '#002b2d',
  greenLight30: '#4dc9c1',
  greenLight90: '#e6f6f7',
  grey: '#a3a6ac',
  greyDark30: '#737479',
  greyDark75: '#29292b',
  greyLight60: '#dbdbde',
  greyLight90: '#f6f6f7',
  grotesqueGreen: '#6af293',
  orange: '#ffa234',
  orangeDark30: '#b37c24',
  orangeDark75: '#40310d',
  orangeLight30: '#ffc870',
  orangeLight90: '#fffaeb',
  pink: '#fe56aa',
  pinkDark30: '#b23e70',
  pinkDark75: '#401624',
  pinkLight30: '#fe8abd',
  pinkLight45: '#fea3c7',
  pinkLight90: '#ffeef3',
  purple: '#be83cf',
  red: '#de350b',
  redDark30: '#9d2f08',
  redDark75: '#381603',
  redLight30: '#e97b54',
  redLight90: '#fcefe7',
  smalt: '#003580',
  smaltDark75: '#000820',
  smaltLight30: '#4d6ca6',
  smaltLight45: '#7389b9',
  violet: '#7622d7',
  white: '#ffffff',
  yellow: '#fadb14',
  yellowDark75: '#3d3f05',
};

const systemPalette = {
  primary: palette.blue,
  hoverPrimary: palette.blueLight30,
  activePrimary: palette.blueDark30,
  primaryBackground: palette.blueLight75,
  primaryLightBackground: palette.blueLight90,
  primaryNeutralBackground: palette.blueDark30,
  primaryDarkBackground: palette.smalt,
  danger: palette.red,
  hoverDanger: palette.redLight30,
  activeDanger: palette.redDark30,
  dangerBackground: palette.redLight90,
  warning: palette.orange,
  hoverWarning: palette.orangeLight30,
  activeWarning: palette.orangeDark30,
  warningBackground: palette.orangeLight90,
  success: palette.green,
  hoverSuccess: palette.greenLight30,
  activeSuccess: palette.greenDark30,
  successBackground: palette.greenLight90,
  error: palette.pink,
  hoverError: palette.pinkLight30,
  activeError: palette.pinkDark30,
  errorBackground: palette.pinkLight90,
  link: palette.blue,
  hoverLink: palette.blueLight30,
  activeLink: palette.blueDark30,
  text: palette.greyDark75,
  lightText: palette.white,
  subduedText: palette.grey,
  defaultBackground: palette.greyLight90,
  defaultLightBackground: palette.white,
  defaultDarkBackground: palette.black,
  neutralBackground: palette.grey,
  neutralDarkBackground: palette.greyDark30,
  defaultBorder: palette.greyLight60,
};

export { palette, systemPalette };
