// hero-design/src/theme/global/typography.ts
interface FontSizes {
  large: number;
  medium: number;
  small: number;
  xlarge: number;
  xsmall: number;
  xxlarge: number;
  xxxlarge: number;
  xxxxlarge: number;
  xxxxxlarge: number;
}

interface FontWeights {
  bold: number;
  light: number;
  regular: number;
  semiBold: number;
}

interface LineHeights {
  large: number;
  medium: number;
  small: number;
  xlarge: number;
  xsmall: number;
  xxlarge: number;
  xxxlarge: number;
  xxxxlarge: number;
  xxxxxlarge: number;
}

const BASE = 16;

const fontSizes: FontSizes = {
  xxxxxlarge: BASE * 2, //   h1
  xxxxlarge: BASE * 1.75, // h2
  xxxlarge: BASE * 1.5, //   h3
  xxlarge: BASE * 1.25, //   h4
  xlarge: BASE * 1.125, //   h5
  large: BASE, //            h6
  medium: BASE * 0.875, //   h7
  small: BASE * 0.75, //     h8
  xsmall: BASE * 0.625, //   h9
};

const fontWeights: FontWeights = {
  light: 200,
  regular: 400,
  semiBold: 600,
  bold: 700,
};

const lineHeights: LineHeights = {
  xxxxxlarge: fontSizes.xxxxxlarge + 8, // h1
  xxxxlarge: fontSizes.xxxxlarge + 8, //   h2
  xxxlarge: fontSizes.xxxlarge + 8, //     h3
  xxlarge: fontSizes.xxlarge + 8, //       h4
  xlarge: fontSizes.xlarge + 8, //         h5
  large: fontSizes.large + 8, //           h6
  medium: fontSizes.medium + 8, //         h7
  small: fontSizes.small + 8, //           h8
  xsmall: fontSizes.xsmall + 8, //         h9
};

export {
  fontSizes,
  fontWeights,
  lineHeights,
  FontSizes,
  FontWeights,
  LineHeights,
};
