// hero-design/src/theme/components/alert.ts
import { fontSizes, fontWeights, lineHeights } from '../global/typography';
import { space } from '../global/space';
import { borderWidths, radii } from '../global/borders';
import { systemPalette } from '../global/colors';

interface AlertSpace {
  compactPadding: string;
  defaultPadding: string;
  iconCompactMarginRight: string;
  iconCompactMarginTop: string;
  iconDefaultMarginRight: string;
  iconDefaultMarginTop: string;
  titleMarginBottom: string;
}

interface AlertFontSizes {
  content: string;
  title: string;
}

interface AlertLineHeights {
  content: string;
  iconCompact: string;
  iconDefault: string;
  title: string;
}

interface AlertFontWeights {
  text: number;
}

interface AlertBorderWidths {
  wrapperFilled: string;
  wrapperOutlined: string;
}

interface AlertRadii {
  wrapper: string;
}

interface AlertColors {
  bgFilledDanger: string;
  bgFilledError: string;
  bgFilledInfo: string;
  bgFilledSuccess: string;
  bgFilledWarning: string;
  bgOutlinedDanger: string;
  bgOutlinedError: string;
  bgOutlinedInfo: string;
  bgOutlinedSuccess: string;
  bgOutlinedWarning: string;
  borderDanger: string;
  borderError: string;
  borderInfo: string;
  borderSuccess: string;
  borderWarning: string;
  textFilled: string;
  textOutlined: string;
}

const alertSpace: AlertSpace = {
  iconDefaultMarginTop: `${space.xxsmall}px`,
  iconDefaultMarginRight: `${space.medium}px`,
  iconCompactMarginTop: `${space.xsmall}px`,
  iconCompactMarginRight: `${space.small}px`,
  defaultPadding: `${space.medium}px`,
  compactPadding: `${space.small}px ${space.medium}px`,
  titleMarginBottom: `${space.xsmall}px`,
};

const alertFontSizes: AlertFontSizes = {
  title: `${fontSizes.large}px`,
  content: `${fontSizes.medium}px`,
};

const alertLineHeights: AlertLineHeights = {
  title: `${lineHeights.large}px`,
  content: `${lineHeights.medium}px`,
  iconDefault: `${lineHeights.large}px`,
  iconCompact: `${lineHeights.medium}px`,
};

const alertFontWeights: AlertFontWeights = {
  text: fontWeights.regular,
};

const alertBorderWidths: AlertBorderWidths = {
  wrapperFilled: `0px`,
  wrapperOutlined: `${borderWidths.base}px`,
};

const alertRadii: AlertRadii = {
  wrapper: `${radii.base}px`,
};

const alertColors: AlertColors = {
  textFilled: systemPalette.lightText,
  textOutlined: systemPalette.text,
  bgFilledSuccess: systemPalette.success,
  bgOutlinedSuccess: systemPalette.successBackground,
  bgFilledInfo: systemPalette.primary,
  bgOutlinedInfo: systemPalette.primaryLightBackground,
  bgFilledWarning: systemPalette.warning,
  bgOutlinedWarning: systemPalette.warningBackground,
  bgFilledDanger: systemPalette.danger,
  bgOutlinedDanger: systemPalette.dangerBackground,
  bgFilledError: systemPalette.error,
  bgOutlinedError: systemPalette.errorBackground,
  borderSuccess: systemPalette.success,
  borderInfo: systemPalette.primary,
  borderWarning: systemPalette.warning,
  borderDanger: systemPalette.danger,
  borderError: systemPalette.error,
};

export {
  alertSpace,
  alertFontSizes,
  alertLineHeights,
  alertFontWeights,
  alertBorderWidths,
  alertRadii,
  alertColors,
};
