// hero-design/src/theme/components/badge.ts
import { space } from '../global/space';
import { fontSizes, fontWeights } from '../global/typography';
import { shadows } from '../global/shadows';
import { sizes } from '../global/sizes';
import { borderWidths } from '../global/borders';
import { systemPalette } from '../global/colors';

interface BadgeColors {
  basicBackground: string;
  basicText: string;
  border: string;
  danger: string;
  error: string;
  primary: string;
  success: string;
  text: string;
  warning: string;
}

interface BadgeFontWeights {
  default: number;
}

interface BadgeFontSizes {
  default: string;
}

interface BadgeShadows {
  default: string;
}

interface BadgeSizes {
  height: string;
  minWidth: string;
}

interface BadgeSpace {
  narrowPadding: string;
  right: number;
  widePadding: string;
}

interface BadgeRadii {
  default: string;
  status: string;
}

interface BadgeBorderWidths {
  default: string;
}

const badgeColors: BadgeColors = {
  basicBackground: systemPalette.defaultLightBackground,
  basicText: systemPalette.subduedText,
  text: systemPalette.lightText,
  border: systemPalette.defaultBorder,
  warning: systemPalette.warning,
  success: systemPalette.success,
  error: systemPalette.error,
  danger: systemPalette.danger,
  primary: systemPalette.primary,
};

const badgeFontWeights: BadgeFontWeights = {
  default: fontWeights.regular,
};

const badgeFontSizes: BadgeFontSizes = {
  default: `${fontSizes.medium}px`,
};

const badgeShadows: BadgeShadows = {
  default: shadows.default,
};

const badgeSizes: BadgeSizes = {
  height: `${sizes.large}px`,
  minWidth: `${sizes.large}px`,
};

const badgeSpace: BadgeSpace = {
  right: 0,
  narrowPadding: `${space.xsmall}px`,
  widePadding: `${space.xsmall}px ${space.small}px`,
};

const badgeRadii: BadgeRadii = {
  default: `${sizes.large / 2}px`,
  status: '50%',
};

const badgeBorderWidths: BadgeBorderWidths = {
  default: `${borderWidths.base}px`,
};

export {
  badgeColors,
  badgeRadii,
  badgeFontSizes,
  badgeFontWeights,
  badgeSpace,
  badgeBorderWidths,
  badgeSizes,
  badgeShadows,
};
