// hero-design/src/theme/components/banner.ts
import { fontSizes, fontWeights, lineHeights } from '../global/typography';
import { space } from '../global/space';
import { systemPalette } from '../global/colors';

interface BannerSpace {
  compactPadding: string;
  defaultPadding: string;
  iconCompactMarginRight: string;
  iconCompactMarginTop: string;
  iconDefaultMarginRight: string;
  iconDefaultMarginTop: string;
  titleMarginBottom: string;
}

interface BannerFontSizes {
  content: string;
  title: string;
}

interface BannerLineHeights {
  content: string;
  iconCompact: string;
  iconDefault: string;
  title: string;
}

interface BannerFontWeights {
  text: number;
}

interface BannerColors {
  bgDanger: string;
  bgError: string;
  bgInfo: string;
  bgSuccess: string;
  bgWarning: string;
}

const bannerSpace: BannerSpace = {
  iconDefaultMarginTop: `${space.xxsmall}px`,
  iconCompactMarginTop: `${space.xsmall}px`,
  iconDefaultMarginRight: `${space.medium}px`,
  iconCompactMarginRight: `${space.small}px`,
  titleMarginBottom: `${space.xsmall}px`,
  defaultPadding: `${space.medium}px`,
  compactPadding: `${space.small}px ${space.medium}px`,
};

const bannerFontSizes: BannerFontSizes = {
  title: `${fontSizes.large}px`,
  content: `${fontSizes.medium}px`,
};

const bannerLineHeights: BannerLineHeights = {
  title: `${lineHeights.large}px`,
  content: `${lineHeights.medium}px`,
  iconDefault: `${lineHeights.large}px`,
  iconCompact: `${lineHeights.medium}px`,
};

const bannerFontWeights: BannerFontWeights = {
  text: fontWeights.regular,
};

const bannerColors: BannerColors = {
  bgInfo: systemPalette.primary,
  bgSuccess: systemPalette.success,
  bgWarning: systemPalette.warning,
  bgDanger: systemPalette.danger,
  bgError: systemPalette.error,
};

export {
  bannerSpace,
  bannerFontSizes,
  bannerLineHeights,
  bannerFontWeights,
  bannerColors,
};
