// hero-design/src/theme/components/breadcrumb.ts
import { fontSizes, fontWeights, lineHeights } from '../global/typography';
import { space } from '../global/space';
import { systemPalette } from '../global/colors';

interface BreadcrumbColors {
  item: string;
  itemActive: string;
}

interface BreadcrumbSpace {
  backIconMarginRight: string;
  separatorMargin: string;
}

interface BreadcrumbFontsizes {
  item: string;
}

interface BreadcrumbLineHeights {
  item: string;
}

interface BreadcrumbFontWeights {
  item: number;
}

const breadcrumbColors: BreadcrumbColors = {
  item: systemPalette.subduedText,
  itemActive: systemPalette.text,
};

const breadcrumbSpace: BreadcrumbSpace = {
  backIconMarginRight: `${space.xsmall}px`,
  separatorMargin: `auto ${space.small}px`,
};

const breadcrumbFontSizes: BreadcrumbFontsizes = {
  item: `${fontSizes.medium}px`,
};

const breadcrumbFontWeights: BreadcrumbFontWeights = {
  item: fontWeights.regular,
};

const breadcrumbLineHeights: BreadcrumbLineHeights = {
  item: `${lineHeights.medium}px`,
};

export {
  breadcrumbColors,
  breadcrumbSpace,
  breadcrumbFontSizes,
  breadcrumbFontWeights,
  breadcrumbLineHeights,
};
