// hero-design/src/theme/components/card.ts
import { fontSizes, lineHeights, fontWeights } from '../global/typography';
import { systemPalette } from '../global/colors';

import { borderWidths, radii } from '../global/borders';
import { space } from '../global/space';

const cardColors = {
  border: systemPalette.defaultBorder,
  background: systemPalette.defaultLightBackground,
  headerBasicBackground: systemPalette.defaultLightBackground,
  headerInfoBackground: systemPalette.primaryBackground,
  headerGreyBackground: systemPalette.defaultBackground,
  headerPrimaryBackground: systemPalette.primary,
  headerPrimaryDarkBackground: systemPalette.primaryDarkBackground,
  headerPrimaryNeutralBackground: systemPalette.primaryNeutralBackground,
  headerDefaultText: systemPalette.text,
  headerPrimaryDarkText: systemPalette.primaryDarkBackground,
  headerLightText: systemPalette.lightText,
  contentText: systemPalette.text,
  contentBorder: systemPalette.defaultBorder,
  contentTextExtra: systemPalette.subduedText,
};

const cardBorderWidths = {
  default: `${borderWidths.base}px`,
};

const cardRadii = {
  default: `${radii.base}px`,
};

const cardSpace = {
  headerMediumPadding: `${space.medium}px`,
  headerSmallPadding: `${space.small}px`,
  headerMargin: `-${borderWidths.base}px -${borderWidths.base}px 0`,
  contentMediumPadding: `${space.medium}px`,
  contentSmallPadding: `${space.small}px`,
  imgMargin: `-${borderWidths.base}px -${borderWidths.base}px 0`,
};

const cardFontWeights = {
  header: fontWeights.semiBold,
  content: fontWeights.regular,
};

const cardFontSizes = {
  header: `${fontSizes.large}px`,
  content: `${fontSizes.medium}px`,
  contentExtra: `${fontSizes.small}px`,
};

const cardLineHeights = {
  header: `${lineHeights.large}px`,
  content: `${lineHeights.medium}px`,
  contentExtra: `${lineHeights.small}px`,
};

export {
  cardColors,
  cardBorderWidths,
  cardRadii,
  cardSpace,
  cardFontWeights,
  cardFontSizes,
  cardLineHeights,
};
