// hero-design/src/theme/components/checkbox.ts
import { fontSizes, lineHeights, fontWeights } from '../global/typography';
import { sizes } from '../global/sizes';
import { borderWidths, radii } from '../global/borders';
import { space } from '../global/space';
import { systemPalette } from '../global/colors';

const checkboxColors = {
  text: systemPalette.text,
  disabledText: systemPalette.subduedText,
  defaultBorder: systemPalette.defaultBorder,
  checkedBorder: systemPalette.primary,
  defaultBackground: systemPalette.defaultLightBackground,
  checkedBackground: systemPalette.primary,
  disabledBackground: systemPalette.defaultBackground,
  defaultCheckmark: systemPalette.lightText,
  disabledCheckmark: systemPalette.subduedText,
  indeterminateBoxBackground: systemPalette.primary,
};

const checkboxSpace = {
  labelPaddingLeft: `${space.large}px`,
  top: `${(lineHeights.medium - sizes.medium) / 2}px`,
  markTop: `${sizes.medium / 2}px`,
  markLeft: `${sizes.medium / 2}px`,
  marginTop: `${sizes.small}px`,
  marginLeft: `${sizes.large}px`,
  indeterminateBoxTop: `${(lineHeights.medium - sizes.small) / 2}px`,
  indeterminateBoxLeft: `${sizes.small / 2}px`,
};

const checkboxSizes = {
  height: `${sizes.medium}px`,
  width: `${sizes.medium}px`,
  markHeight: `${sizes.small}px`,
  markWidth: `${sizes.xsmall}px`,
  indeterminateBoxHeight: `${sizes.small}px`,
  indeterminateBoxWidth: `${sizes.small}px`,
};

const checkboxBorderWidths = {
  default: `${borderWidths.base}px`,
};

const checkboxRadii = {
  default: `${radii.base}px`,
  indeterminateBox: `${radii.base / 2}px`,
};

const checkboxFontSizes = {
  default: `${fontSizes.medium}px`,
};

const checkboxLineHeights = {
  default: `${lineHeights.medium}px`,
};

const checkboxFontWeights = {
  default: fontWeights.regular,
};

export {
  checkboxColors,
  checkboxSpace,
  checkboxSizes,
  checkboxBorderWidths,
  checkboxRadii,
  checkboxFontSizes,
  checkboxFontWeights,
  checkboxLineHeights,
};
