// hero-design/src/theme/components/notification.ts
import { fontSizes, lineHeights } from '../global/typography';
import { shadows } from '../global/shadows';
import { radii } from '../global/borders';
import { space } from '../global/space';
import { systemPalette } from '../global/colors';

const notificationSpace = {
  iconMarginTop: `${space.xxsmall}px`,
  iconMarginRight: `${space.medium}px`,
  padding: `${space.medium}px`,
  titleMarginBottom: `${space.xsmall}px`,
};

const notificationFontSizes = {
  title: `${fontSizes.large}px`,
  content: `${fontSizes.medium}px`,
};

const notificationLineHeights = {
  title: `${lineHeights.large}px`,
  content: `${lineHeights.medium}px`,
  icon: `${lineHeights.large}px`,
};

const notificationRadii = {
  default: `${radii.base}px`,
};

const notificationShadows = {
  default: shadows.default,
};

const notificationColors = {
  text: systemPalette.text,
  bg: systemPalette.defaultLightBackground,
};

export {
  notificationSpace,
  notificationFontSizes,
  notificationLineHeights,
  notificationRadii,
  notificationColors,
  notificationShadows,
};
