// hero-design/src/theme/components/progress.ts
import { fontSizes, fontWeights, lineHeights } from '../global/typography';
import { sizes } from '../global/sizes';
import { space } from '../global/space';
import { radii } from '../global/borders';
import { systemPalette } from '../global/colors';

interface ProgressColors {
  background: string;
  circleStroke: string;
  danger: string;
  primary: string;
  success: string;
  text: string;
}

interface ProgressFontSizes {
  barMedium: string;
  barSmall: string;
  circleMedium: string;
  circleSmall: string;
}

interface ProgressFontWeights {
  default: number;
}

interface ProgressLineHeights {
  barMedium: string;
  barSmall: string;
}

interface ProgressSpace {
  barTextPadding: string;
}

interface ProgressSizes {
  barMediumHeight: string;
  barSmallHeight: string;
  barTextWidth: string;
  barWrapperMediumHeight: string;
  barWrapperSmallHeight: string;
  circleMedium: string;
  circleMediumStroke: string;
  circleSmall: string;
  circleSmallStroke: string;
}

interface ProgressRadii {
  bar: string;
}

const progressColors: ProgressColors = {
  text: systemPalette.subduedText,
  background: systemPalette.defaultBackground,
  success: systemPalette.success,
  danger: systemPalette.danger,
  primary: systemPalette.primary,
  circleStroke: systemPalette.defaultBackground,
};

const progressFontSizes: ProgressFontSizes = {
  barSmall: `${fontSizes.xsmall}px`,
  barMedium: `${fontSizes.small}px`,
  circleSmall: `${fontSizes.large}px`,
  circleMedium: `${fontSizes.xxxlarge}px`,
};

const progressFontWeights: ProgressFontWeights = {
  default: fontWeights.regular,
};

const progressLineHeights: ProgressLineHeights = {
  barSmall: `${lineHeights.xsmall}px`,
  barMedium: `${lineHeights.small}px`,
};

const progressSpace: ProgressSpace = {
  barTextPadding: `${space.small}px`,
};

const progressSizes: ProgressSizes = {
  barWrapperSmallHeight: progressLineHeights.barSmall,
  barWrapperMediumHeight: progressLineHeights.barMedium,
  barSmallHeight: `${sizes.xsmall}px`,
  barMediumHeight: `${sizes.small}px`,
  barTextWidth: `${sizes.xlarge}px`,
  circleSmall: '64px',
  circleMedium: '112px',
  circleSmallStroke: `${sizes.xsmall}px`,
  circleMediumStroke: `${sizes.small}px`,
};

const progressRadii: ProgressRadii = {
  bar: `${radii.base}px`,
};

export {
  progressColors,
  progressFontSizes,
  progressFontWeights,
  progressLineHeights,
  progressSpace,
  progressSizes,
  progressRadii,
};
