// hero-design/src/theme/components/radio.ts
import { sizes } from '../global/sizes';
import { fontSizes, lineHeights, fontWeights } from '../global/typography';
import { borderWidths, radii } from '../global/borders';
import { space } from '../global/space';
import { systemPalette } from '../global/colors';

const radioSizes = {
  height: `${sizes.medium}px`,
  width: `${sizes.medium}px`,
  checkmarkHeight: `${sizes.small}px`,
  checkmarkWidth: `${sizes.small}px`,
  buttonSmall: `${sizes.large}px`,
  buttonMedium: `${sizes.xlarge}px`,
  buttonLarge: `${sizes.xxxlarge}px`,
};

const radioColors = {
  defaultBorder: systemPalette.defaultBorder,
  disabledBorder: systemPalette.defaultBorder,
  hoverBorder: systemPalette.primary,
  activeBorder: systemPalette.primary,
  defaultText: systemPalette.text,
  hoverText: systemPalette.primary,
  disabledText: systemPalette.subduedText,
  disabledSelectedText: systemPalette.lightText,
  filledSelectedText: systemPalette.lightText,
  defaultCheckmark: systemPalette.primary,
  disabledCheckmark: systemPalette.neutralBackground,
  disabledBackground: systemPalette.defaultBackground,
  buttonDisabledBackground: systemPalette.defaultBackground,
  buttonCheckedBackground: systemPalette.primary,
  buttonDisabledCheckedBackground: systemPalette.neutralBackground,
};

const radioRadii = {
  default: '50%',
  button: `${radii.base}px`,
};

const radioBorderWidths = {
  default: `${borderWidths.base}px`,
};

const radioSpace = {
  margin: `0 ${space.small}px 0 0`,
  groupVerticalMargin: `${space.small}px`,
  groupHorizontalMargin: `${space.medium}px`,
  buttonSmallPadding: `${space.xsmall}px ${space.small}px`,
  buttonMediumPadding: `${space.small}px ${space.medium}px`,
  buttonLargePadding: `${space.small}px ${space.medium}px`,
};

const radioFontSizes = {
  default: `${fontSizes.medium}px`,
  buttonSmall: `${fontSizes.small}px`,
  buttonMedium: `${fontSizes.medium}px`,
  buttonLarge: `${fontSizes.xlarge}px`,
};

const radioLineHeights = {
  default: `${lineHeights.medium}px`,
};

const radioFontWeights = {
  default: fontWeights.regular,
};

export {
  radioSizes,
  radioColors,
  radioRadii,
  radioBorderWidths,
  radioSpace,
  radioFontSizes,
  radioLineHeights,
  radioFontWeights,
};
