// hero-design/src/theme/components/statistic.ts
import { space } from '../global/space';
import { fontSizes, lineHeights, fontWeights } from '../global/typography';
import { systemPalette } from '../global/colors';

interface StatisticColors {
  contentText: string;
  contentWhite: string;
  titleText: string;
  titleWhite: string;
}

interface StatisticFontWeights {
  content: number;
}

interface StatisticFontSizes {
  content: string;
  title: string;
}

interface StatisticLineHeights {
  default: string;
}

interface StatisticSpace {
  prefixWrapperMargin: string;
  suffixWrapperMargin: string;
  titleWrapperMargin: string;
}

const statisticColors: StatisticColors = {
  titleWhite: systemPalette.lightText,
  titleText: systemPalette.subduedText,
  contentWhite: systemPalette.lightText,
  contentText: systemPalette.text,
};

const statisticFontWeights: StatisticFontWeights = {
  content: fontWeights.semiBold,
};

const statisticFontSizes: StatisticFontSizes = {
  title: `${fontSizes.medium}px`,
  content: `${fontSizes.xxxlarge}px`,
};

const statisticLineHeights: StatisticLineHeights = {
  default: `${lineHeights.medium}px`,
};

const statisticSpace: StatisticSpace = {
  titleWrapperMargin: `0 0 ${space.small}px 0`,
  prefixWrapperMargin: `0 ${space.xsmall}px 0 0`,
  suffixWrapperMargin: `0 0 0 ${space.xsmall}px`,
};

export {
  statisticColors,
  statisticFontWeights,
  statisticFontSizes,
  statisticLineHeights,
  statisticSpace,
};
