// hero-design/src/theme/components/table.ts
import { fontSizes, lineHeights, fontWeights } from '../global/typography';
import { borderWidths, radii } from '../global/borders';
import { space } from '../global/space';
import { sizes } from '../global/sizes';
import { systemPalette } from '../global/colors';

const tableSizes = {
  expandableIcon: `${sizes.medium}px`,
  itemsPerPageInputWidth: '65px',
};

const tableColors = {
  background: systemPalette.defaultLightBackground,
  headerText: systemPalette.text,
  headerBackground: systemPalette.defaultBackground,
  rowFocusBackground: systemPalette.primaryLightBackground,
  cellText: systemPalette.text,
  cellBorder: systemPalette.defaultBorder,
  expandableIconBorder: systemPalette.defaultBorder,
};

const tableSpace = {
  headerPadding: `${space.medium}px`,
  cellPadding: `${space.medium}px`,
  sortingIconMarginLeft: `${space.small}px`,
  expandableIconPadding: `${space.xxsmall}px`,
  paginationMarginTop: `${space.medium}px`,
};

const tableFontSizes = {
  header: `${fontSizes.medium}px`,
  cell: `${fontSizes.medium}px`,
  sortingIcon: `${fontSizes.xsmall}px`,
  expandableIcon: `${fontSizes.xsmall}px`,
};

const tableFontWeights = {
  header: fontWeights.semiBold,
  cell: fontWeights.regular,
};

const tableLineHeights = {
  header: `${lineHeights.medium}px`,
  cell: `${lineHeights.medium}px`,
  sortingIcon: `${lineHeights.xsmall}px`,
};

const tableBorderWidths = {
  cell: `${borderWidths.base}px`,
  expandableIcon: `${borderWidths.base}px`,
};

const tableRadii = {
  expandableIcon: `${radii.base}px`,
};

export {
  tableSizes,
  tableColors,
  tableSpace,
  tableFontSizes,
  tableFontWeights,
  tableLineHeights,
  tableBorderWidths,
  tableRadii,
};
